### ISSUE TITLE
Nondeterministic java.lang.StackOverflowError when compiling 

### ISSUE THREAD

**Issue by mrubin** [sbt/sbt#5299]
## steps
sbt version: 1.2.8

## problem

```
[error] ## Exception when compiling 846 sources to /var/lib/jenkins/workspace/my-project/target/scala-2.12/classes
[error] null
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error] scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4839)
[error] scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4867)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5670)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error] scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4839)
[error] scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4867)
...
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error] scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4839)
[error] scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4867)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5670)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error]            
[error] java.lang.StackOverflowError
[error]   at scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error]   at scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error]   at scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error]   at scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
...
```

## notes

The issue is non-deterministic. When it occurs, we do a `clean` followed by attempting to recompile and after some number of failed attempts (with the same error), the build will start working again.

We have already increased the stack size. Running `sbt -v` shows the following:

```
java
-Xms1024m
-Xmx4096m
-Xss8m
-jar
/usr/share/sbt/bin/sbt-launch.jar
```

The behavior of the issue has not changed.

**Comment by eed3si9n** [comment]
@mrubin Thanks for the report.

1. Could you try the latest sbt 1.3.4?
2. Could you fill in the reproduction steps? It doesn't have to be minimized or takes 1000 tries to make it happen, but without zero information on how to reproduce it on our machine, there's no chance we can actually work on this issue.

**Comment by mrubin** [comment]
Thank you @eed3si9n .
1. We will try sbt 1.3.4.
2. I understand the need for reproduction steps. Unfortunately this is a large internal project that cannot be shared and the issue does not happen to us consistently either. Are there any additional debug flags we could add to the compilation flags such that when the issue occurs, we can pass some info back to you?

Pasting from our Jenkins build log, you can see that there were 9 failures in a row followed by a success. There were no code changes in between... when this happens, we just trigger the Jenkins job over and over until it succeeds.

```
Success > Console Output#3​44​9 Dec 12, 2019 9:06 AM 
Failed > Console Output#3​44​8 Dec 12, 2019 8:29 AM 
Failed > Console Output#3​44​7 Dec 12, 2019 8:27 AM 
Failed > Console Output#3​44​6 Dec 12, 2019 8:24 AM 
Failed > Console Output#3​44​5 Dec 12, 2019 8:22 AM 
Failed > Console Output#3​44​4 Dec 11, 2019 8:25 PM 
Failed > Console Output#3​44​3 Dec 11, 2019 3:28 PM 
Failed > Console Output#3​44​2 Dec 11, 2019 3:26 PM 
Failed > Console Output#3​44​1 Dec 11, 2019 3:23 PM 
Failed > Console Output#3​44​0 Dec 11, 2019 3:13 PM 
Success > Console Output#3​43​9 Dec 11, 2019 1:34 PM 
Failed > Console Output#3​43​8 Dec 11, 2019 1:31 PM 
Success > Console Output#3​43​7 Dec 11, 2019 6:29 AM 
Failed > Console Output#3​43​6 Dec 11, 2019 4:53 AM 
Success > Console Output#3​43​5 Dec 10, 2019 5:28 PM 
Failed > Console Output#3​43​4 Dec 10, 2019 5:25 PM 
Failed > Console Output#3​43​3 Dec 10, 2019 2:08 PM 
Success > Console Output#3​43​2 Dec 10, 2019 11:38 AM 
Success > Console Output#3​43​1 Dec 10, 2019 11:35 AM 
Success > Console Output#3​43​0 Dec 10, 2019 11:23 AM 
Failed > Console Output#3​42​9 Dec 10, 2019 11:19 AM 
Success > Console Output#3​42​8 Dec 10, 2019 9:57 AM 
Success > Console Output#3​42​7 Dec 10, 2019 9:52 AM 
Failed > Console Output#3​42​6 Dec 10, 2019 8:27 AM 
Failed > Console Output#3​42​5 Dec 10, 2019 8:03 AM 
Success > Console Output#3​42​4 Dec 9, 2019 6:34 AM 
Success > Console Output#3​42​3 Dec 9, 2019 6:31 AM 
Success > Console Output#3​42​2 Dec 7, 2019 1:03 AM 
Success > Console Output#3​42​1 Dec 6, 2019 11:46 AM
```

**Comment by mrubin** [comment]
@eed3si9n - the issue is still occurring (sporadically) with sbt 1.3.4. A snippet of the output is below.

Are there any debug compiler flags we could use which might shed additional light on the issue, next time it happens?

```
[error] ## Exception when compiling 852 sources to /var/lib/jenkins/workspace/my-project/target/scala-2.12/classes
[error] java.lang.StackOverflowError
[error] scala.tools.nsc.typechecker.AnalyzerPlugins.pluginsPt$(AnalyzerPlugins.scala:329)
[error] scala.tools.nsc.transform.Erasure.pluginsPt(Erasure.scala:22)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5700)
[error] scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error] scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4839)
[error] scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4867)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5670)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
[error] scala.tools.nsc.typechecker.Typers$Typer.silent(Typers.scala:714)
[error] scala.tools.nsc.typechecker.Typers$Typer.normalTypedApply$1(Typers.scala:4839)
[error] scala.tools.nsc.typechecker.Typers$Typer.typedApply$1(Typers.scala:4867)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed1(Typers.scala:5670)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.transform.Erasure$Eraser.adaptMember(Erasure.scala:734)
[error] scala.tools.nsc.transform.Erasure$Eraser.typed1(Erasure.scala:809)
[error] scala.tools.nsc.typechecker.Typers$Typer.typed(Typers.scala:5716)
[error] scala.tools.nsc.typechecker.Typers$Typer.$anonfun$typed1$39(Typers.scala:4837)
```

**Comment by eed3si9n** [comment]
I am not sure. If it's happening nondetermistically, you are actually just running out of stack.
Try increasing it via `-Xss5m` etc. Otherwise please report to scala/bug.

**Comment by mrubinwch** [comment]
We are already using `-Xss8m` and that has not helped.

**Comment by MrGigabit** [comment]
I was running into a similar issue. I had increased the max heap size to 3g `Xmx3072M`. Still throwing a stackoverflow. 

I updated my minimum heap size to 1g `Xss1024M`. This allowed me to compile. 

Note, I was using Scala plugin in Intellij, and the bundled sbt compiler. You can update min heap by passing a command line argument to the VM from: `Preferences > Build, Execution, Deployment > Build Tools > sbt`

![image](https://user-images.githubusercontent.com/31316550/120848146-6c466c00-c542-11eb-8a05-879e025b3bc7.png)

If you are not using the Scala plugin in Intellij but a stand alone installation, you can update your min and max heap size in your `sbtopts` file located in `/usr/local/etc/sbtopts` (this is on mac)

**Comment by eed3si9n** [comment]
1GB heap size sounds like a bug of a compiler. See for example https://github.com/scala/bug/issues/9310, which was fixed in Scala 2.13.0.

**Comment by MrGigabit** [comment]
Odd....I am using Scala 2.13.6, specified in my `build.sbt`

`scalaVersion := "2.13.6"`
